core/src/main/scala/kafka/server/AlterIsrManager.scala

```diff
@@ -165,9 +165,19 @@ class DefaultAlterIsrManager(
       new ControllerRequestCompletionHandler {
         override def onComplete(response: ClientResponse): Unit = {
           debug(s"Received AlterIsr response $response")
-          val body = response.responseBody().asInstanceOf[AlterIsrResponse]
           val error = try {
-            handleAlterIsrResponse(body, message.brokerEpoch, inflightAlterIsrItems)
+            if (response.authenticationException != null) {
+              // For now we treat authentication errors as retriable. We use the
+              // `NETWORK_EXCEPTION` error code for lack of a good alternative.
+              // Note that `BrokerToControllerChannelManager` will still log the
+              // authentication errors so that users have a chance to fix the problem.
+              Errors.NETWORK_EXCEPTION
+            } else if (response.versionMismatch != null) {
```

Comment:
Seems like it's pretty common t check for these two fields. Could we have an enum to represent the two errors and then have a function from the enum the Error? Seems a bit easier to use and less error prone.

Reply:
It's a good suggestion. I think what we want is for `ClientResponse` to expose something like `Either[AbstractResponse, Errors]`. This would prevent bugs like this in the future since it forces the code to unwrap the response. I'm willing to give it a shot here, but I suspect it will cause a much bigger diff. My intent here was to try and keep the PR small so that we could get it into 2.8. Let me know what you think.

Reply:
I'm fine to improve this post 2.8, let's file a JIRA please.

Reply:
Filed this: https://issues.apache.org/jira/browse/KAFKA-12353